asab/web/authn/oauth/proxy.py

```diff
@@ -0,0 +1,136 @@
+import abc
+import aiohttp
+import logging
+
+from ...rest import json_response
+
+#
+
+L = logging.getLogger(__name__)
+
+#
+
+
+class ABCOAuthProxy(abc.ABC):
+
+	@abc.abstractmethod
+	def get_oauth_server_id(self):
+		pass
+
+	@abc.abstractmethod
+	def get_token_url(self):
+		pass
+
+	@abc.abstractmethod
+	def get_identity_url(self):
+		pass
+
+	@abc.abstractmethod
+	def get_invalidate_url(self):
+		pass
+
+	@abc.abstractmethod
+	def get_forgot_url(self):
+		pass
+
+
+class GitHubOAuthProxy(ABCOAuthProxy):
+
+	def get_oauth_server_id(self):
+		return "github.com"
+
+	def get_token_url(self):
+		return "https://github.com/login/oauth/access_token"
+
+	def get_identity_url(self):
+		return "https://api.github.com/user"
+
+	def get_invalidate_url(self):
+		return None
+
+	def get_forgot_url(self):
+		return None
+
+
+def oauthclient_proxy_factory(*args, container, proxies, **kwargs):
+	"""
+	Serves to add proxy endpoints to the provided web server container,
+	so the client applications may get/post information about the OAuth login.
+
+	The proxied endpoints include: POST token, GET identity, POST invalidate and POST forgot.
+	Every OAuth 2.0 server is required to implement token and identity endpoints, while invalidate and forgot are optional.
+	"""
+
+	proxies_dict = {}
+	for proxy in proxies:
+		proxies_dict[proxy.get_oauth_server_id()] = proxy
+
+	# POST -> to receive access token and refresh token
+	async def token(request):
```

Comment:
The architecture is incorrect ... these handler method (GET, POST) should go into OAuthProxy() object, that is initiated with a specific configuration of the provider.

add_oauth_client() installs these methods into a container.

Reply:
@ateska Please see the latest changes. I can do it in a way that the add_oauth_client() inserts proxy class methods DIRECTLY into the container, but it would mean, there could be only one proxy class at one time (so let us say, only a GitHub proxy, not both TeskaLabs and GitHub).

Reply:
@ateska So the method will be refactored to a class OAuthForwarder and it will obtain all proxies (OAuthProxy) as it is now via a parameter proxies=[]? Can you be please more specific on how it should look from the end-user perspective?

Reply:
Like this?

```
		asab.web.authn.oauth.OAuthForwarder(container=container, providers=[asab.web.authn.oauth.OAuthProvider(config={
				"oauth_server_id": "github.com",
				"token_url": "https://github.com/login/oauth/access_token",
				"identity_url": "https://api.github.com/user",
			})])
```